src/pending_connection/connect.rs

```diff
@@ -1,208 +1,261 @@
-    local_sockid: SocketID,
     remote: SocketAddr,
-) -> Result<(i32, HandshakeControlInfo), Error>
+    local_sockid: SocketID,
+    local_addr: IpAddr,
+    tsbpd_latency: Duration,
+    _crypto: Option<(u8, String)>,
+) -> Result<Connection, Error>
 where
     T: Stream<Item = Result<(Packet, SocketAddr), Error>>
         + Sink<(Packet, SocketAddr), Error = Error>
         + Unpin,
 {
-    let mut send_interval = interval(Duration::from_millis(100));
-
-    info!("Connecting to {}...", remote);
-
-    let request_packet = Packet::Control(ControlPacket {
-        dest_sockid: SocketID(0),
-        timestamp: 0, // TODO: this is not zero in the reference implementation
-        control_type: ControlTypes::Handshake(HandshakeControlInfo {
-            init_seq_num: rand::random(),
-            max_packet_size: 1500, // TODO: take as a parameter
-            max_flow_size: 8192,   // TODO: take as a parameter
-            socket_id: local_sockid,
-            shake_type: ShakeType::Induction,
-            peer_addr: local_addr,
-            syn_cookie: 0,
-            info: HandshakeVSInfo::V4(SocketType::Datagram),
-        }),
-    });
-
-    sock.send((request_packet.clone(), remote)).await?;
-
-    info!("Sent first packet to {}", remote);
-
-    // get first response packet
-    let (timestamp, hs_info) = loop {
-        // just drop the future that didn't finish first
-        let (packet, addr) = select! {
-            _ = send_interval.tick().fuse() => {sock.send((request_packet.clone(), remote)).await?; continue},
-            res = get_packet(sock).fuse() => res?
+    let mut connect = Connect(
+        ConnectConfiguration {
+            remote,
+            local_sockid,
+            local_addr,
+            tsbpd_latency,
+        },
+        ConnectState::new(),
+    );
+
+    let mut tick_interval = interval(Duration::from_millis(100));
+    loop {
+        let result = select! {
+            now = tick_interval.tick().fuse() => connect.next_tick(now),
+            packet = get_packet(sock).fuse() => connect.next_packet(packet?),
         };
 
-        debug!("Got packet from {}", addr);
-        // make sure the socket id and packet type match
-        if let Packet::Control(ControlPacket {
-            timestamp,
-            control_type: ControlTypes::Handshake(info @ HandshakeControlInfo { .. }),
-            ..
-        }) = packet
-        {
-            if info.shake_type != ShakeType::Induction {
-                info!(
-                    "Expected Induction (1) packet, got {:?} ({})",
-                    info.shake_type, info.shake_type as u32
-                );
-                continue;
+        match result {
+            Ok(Some(packet)) => {
+                sock.send(packet).await?;
             }
-            if addr != remote {
-                warn!("Expected packet from {}, got {}", remote, addr);
-                continue;
+            Err(e) => {
+                eprintln!("{:?}", e);
```

Comment:
Don't like this, I would much prefer you use the log macros instead so it's configurable.